### DISCUSSION TITLE
Native JSONB support for SQLite

### DISCUSSION THREAD

**Discussion by TapGhoul** [SeaQL/sea-orm#2426] [Ideas]
Heyo! Ever since SQLite 3.45.0, there is native support for JSONB on-disk and can be switched in and out on the fly to textual JSON. If using sqlite-side JSON ops, it performs faster than standard JSON, which is nice for raw queries.

Note, the actual format isn't really any smaller than regular JSON for most data that is to be held. It is smaller for a few primitives (null, true and false each only take 1 byte) and most values - even ints and floats - encode the same as standard JSON. This just moves the payload up to the start of the value on-disk.
In cases where the byte (ascii) size of the value is less than 11 bytes in length, it will be smaller - only a single byte header. But in cases where it's 12 or more, it has no actual size gains.

It can be transparently upgraded in the case of sqlite, as the function `json(x)` can take both JSON and JSONB internally, and same with the `jsonb(x)` function to turn it into jsonb itself, so it'd be a matter of wrapping any inserts in `jsonb()` and any loads in `json()` - so adding this imposes no breakages to existing on-disk data, with the exception that raw queries sitting outside of the ORM/query builder that are expecting text may no longer function as expected.

There is no way to define this as a custom column type here that has a default affinity of BLOB, but thanks to sqlite's loose typing rules you can reuse the same type - or just do a custom one (`jsonb_blob` or something) and it'll sort itself out, as per the column/data affinity rules.

Information can be found here:
https://sqlite.org/jsonb.html
https://sqlite.org/json1.html#jminib

**Comment by doveq** [comment]
how do i use jsonb() when insert data?  

like this
ActiveModel {
   data: Set( jsonb('{"a": 123}') )
}

**Comment by JMLX42** [comment]
As the author of Diesel's SQLite JSONB support ([diesel-rs/diesel#4284](https://github.com/diesel-rs/diesel/pull/4284)), I ran into this same gap from the SeaORM side on a production codebase and want to share a recipe that works end-to-end today — and answer @doveq's question about using `jsonb()` on insert.

**TL;DR:** there's still no native SQLite-JSONB-at-rest in SeaORM (nor in sqlx — see [launchbadge/sqlx#4101](https://github.com/launchbadge/sqlx/issues/4101)), but you can get it without touching entity codegen, exactly along the loose-typing lines the OP anticipated.

## The approach

**1. Declare the column with TEXT affinity, not BLOB.** SeaORM has no BLOB-default custom column type — but you don't need one. A `jsonb()` value written into a `varchar`/TEXT-affinity column is *still stored as a BLOB*: SQLite's TEXT affinity only coerces **numeric** values to text and leaves BLOBs untouched. So the column is JSONB at rest (`typeof(col) = 'blob'`, verified) while the generated entity field stays `String` / `serde_json::Value` — no codegen drift, no `Vec<u8>` fields to convert.

```rust
// migration
ColumnDef::new(Packet::Properties)
    .string()                                        // TEXT affinity => entity stays String
    .not_null()
    .check(Expr::cust("json_valid(properties, 8)"))  // flag 8 = strict JSONB
```

**2. Wrap `json()` on read and `jsonb()` on write at the query layer** (this is the `jsonb()`-on-insert answer). SeaORM's per-column `select_as` / `save_as` are CAST-only, so the wrapping lives in the insert/select expressions, not the entity:

```rust
// write: bind jsonb(?) instead of a plain ActiveValue::Set(text)
Func::cust(Alias::new("jsonb")).arg(Expr::val(serialized_json))   // Some(text) => jsonb(?)
// ...and bind a plain SQL NULL for None — never jsonb('null')

// read: project json(col) back to text so the String / serde_json round-trip is unchanged
Func::cust(Alias::new("json")).arg(Expr::col(Column::Properties))
```

Since an insert needs a per-column expression and `Insert`/`ActiveModel` can't carry one, the insert becomes a hand-built `sea_query` INSERT (then re-select to return the model); the update path can use `update_many().col_expr(col, jsonb_expr)`.

## A gotcha that isn't in any docs

If you filter these columns with the JSON/string/array **functions** (not just `=`), the query builder guards them with `json_valid(col)` — the **single-argument, text-only** form, which returns `0` for a JSONB blob and silently drops every row (the guarded `CASE` collapses to `NULL`). The fix is `json_valid(col, 9)` — flag `0x01 | 0x08` = "canonical JSON text **or** strict JSONB" — which keeps still-text columns working too. Plain comparison operators (`=`, `<`, …) are unaffected; only the function-filter guards need it.

## Why do this rather than a native binary type

Diesel's SQLite `Jsonb` (#4284) decodes the binary format byte-for-byte. That's transparent, but — as Diesel's own docs [now warn](https://docs.rs/diesel/latest/diesel/sql_types/struct.Jsonb.html) — it's coupled to a format SQLite documents as *internal*, so future SQLite releases "might break" it. The SQL-function approach above never parses the bytes itself; it only asks SQLite to `json()` / `jsonb()`, so it rides whatever the installed SQLite does — and for SeaORM specifically it needs zero changes to entity generation.

A first-class `ColumnType` that maps to `serde_json::Value` and emits `jsonb()` / `json()` automatically would obviously be nicer, but the recipe above unblocks it today.

<sub>Researched and drafted with the help of an AI assistant (Claude Code); the recipe is distilled from a working production implementation and its passing tests.</sub>

**Reply by TapGhoul** [reply]
@JMLX42 

Ignoring the fact that this reads as AI slop, regardless of the docs at the bottom - the point of jsonb in sqlite is to allow for using values within JSON fields as part of SQL queries, and being able to do *that* faster than the json type.

jsonb is a blob under the hood in regards to its affinity, it's json but using prefix lengths and such to do jumps, but is otherwise the same size (as documented on sqlite's own docs) but it allows you to do ad-hoc queries using the -> and ->> operators (with higher perf in the b variant), for example:

```sql
SELECT message
  FROM logs l
  INNER JOIN origins o
    ON l.rawJson ->> '$.origin' = o.id
  LIMIT 50;
```

Which when doing unexpected analytics queries that aren't a part of the ORM (a huge reason to use types natively supported by the DB engine here!) making these transparent is useful so they actually get used when doing things like answering business questions.

Your response talks about inserting, which the above can do, but it doesn't help with querying and returning JSONB for use with serde etc. Of course you can do ad-hoc translation, but the point here is natively/transparently translating these - as an ORM ought to given it's a natively supported feature - and so your answer doesn't really help here.

I wish you, along with many others, wouldn't use an LLM to generate an answer without actually reading the initial question properly, or if you must, at least ensure that the LLM actually answers the problem properly. Drive-by LLM responses only cause problems for people. Your answer doesn't even answer @doveq's question properly here, because they want an ActiveModel, and your info ignores a large amount of what ActiveModel actually does.

If you respond to this, please resist the temptation to use an LLM to answer for you, and actually do the research needed to answer the question properly. Github is a great place to collaborate and help others, and the discussion pages are a great place to do that with actual people. If they wanted to ask an LLM, they'd just ask an LLM - there's enough free trials and free access of various models that it's actively insulting to assume that the person asking hasn't at the very least performed a quick search for themselves before coming and asking other people. Assume better of other people, we are all real people behind the screen (at least I'd hope).

**Reply by JMLX42** [reply]
@TapGhoul point by point, because the scope you're describing isn't the scope this discussion established.

> the point of jsonb in sqlite is to allow for using values within JSON fields as part of SQL queries […] it allows you to do ad-hoc queries using the `->` and `->>` operators

**1. That is exactly what the recipe delivers — and it's the only thing that can.** `->>` gets the JSONB fast path if and only if the bytes on disk are JSONB. Today SeaORM writes TEXT, so your analytics query pays a full parse on every row. The whole point of the recipe is `typeof(col) = 'blob'` — JSONB *at rest*, verified. Once that holds, your `l.rawJson ->> '$.origin' = o.id` works, with the b-variant performance, unchanged, with no ORM involvement at all. You describe those queries yourself as "not a part of the ORM" — a query outside the ORM needs no ORM feature; it needs the storage format. That's the deliverable.

**2. `->`/`->>` and "ad-hoc analytics" appear nowhere in your opening post.** What your OP actually specifies is: *"it'd be a matter of wrapping any inserts in `jsonb()` and any loads in `json()`"*, plus *"There is no way to define this as a custom column type here that has a default affinity of BLOB, but thanks to sqlite's loose typing rules you can reuse the same type."* My comment is the mechanism you proposed, resolved down to the two questions you left open:

- your BLOB-affinity blocker is a non-issue — TEXT affinity coerces only *numeric* values, so a `jsonb()` value in a `varchar` column stays a BLOB. No custom column type needed, and the generated entity field stays `String` / `serde_json::Value` with zero codegen drift.
- your own compatibility caveat — *"raw queries sitting outside of the ORM […] expecting text may no longer function as expected"* — has a concrete, undocumented instance I hit and reported: the query builder's function-filter guard emits single-argument `json_valid(col)`, which returns `0` for a JSONB blob, so the guarded `CASE` collapses to `NULL` and **silently drops every row**. Fix: `json_valid(col, 9)` (`0x01|0x08` = canonical JSON text *or* strict JSONB), which keeps still-text columns working during migration. That is the failure mode you predicted in 2024, diagnosed and fixed. If you have a source that documents it, I'd like to see it.

**3. "Your response talks about inserting" — because that was the open question.** The only message between your OP and mine is @doveq asking, verbatim: *"how do i use jsonb() when insert data? like this ActiveModel { data: Set( jsonb('{"a": 123}') ) }"*. Answering the question that was asked is not missing the point.

**4. And it answers it in ActiveModel terms specifically.** The reason `Set(jsonb(...))` cannot work is named explicitly: `ActiveValue::Set` carries a *value*, not an expression, and SeaORM's per-column `select_as` / `save_as` are CAST-only — so the wrapping has to live in the statement. Hence: hand-built `sea_query` INSERT then re-select, or the update-path alternative `update_many().col_expr(col, jsonb_expr)`, plus the `None` → plain SQL `NULL` trap (never `jsonb('null')`). If "ignores a large amount of what ActiveModel actually does" points at something more specific, name it and I'll address it.

**5. "doesn't help with […] returning JSONB for use with serde" — second code block of that section.** `Func::cust("json").arg(Expr::col(Column::Properties))` on read projects back to text so the `String` / `serde_json` round-trip is byte-identical. Read and write are both there.

**6. On "actually do the research" and native support.** My comment doesn't claim to close this issue — it says, in as many words: *"A first-class `ColumnType` that maps to `serde_json::Value` and emits `jsonb()` / `json()` automatically would obviously be nicer, but the recipe above unblocks it today."* You're arguing against a position I already conceded. It also hands the maintainers the one design decision that feature turns on, from the only prior art that exists: Diesel's SQLite `Jsonb` ([#4284](https://github.com/diesel-rs/diesel/pull/4284)) decodes the binary format byte-for-byte, which is why Diesel's own docs now warn it is coupled to a format SQLite documents as *internal* and that future releases "might break" it. The `json()` / `jsonb()` route never parses the bytes and rides whatever SQLite is installed. That trade-off is the blueprint for doing this natively in SeaORM.

**7. On "AI slop".** The use of AI assistance was disclosed in the comment, unprompted. Slop is content whose claims don't survive checking. Every claim in mine is checkable, and cited: SQLite's `json_valid` flag semantics, Diesel's own drift warning, [sqlx#4101](https://github.com/launchbadge/sqlx/issues/4101) for the parallel gap, [diesel#4284](https://github.com/diesel-rs/diesel/pull/4284) for the implementation — and the recipe is distilled from a production implementation with passing tests, not from a chat window. Judge it on whether it is *wrong*, not on how it was drafted. So far, every specific technical objection raised is answered by paragraphs already in the comment.

**Reply by TapGhoul** [reply]
And you're still using AI to respond, so I take it you can't be bothered reading what I say. What are you doing. If you have something worthwhile to say you shouldn't be putting it through an LLM.

While I don't care to read all of your reply, since you didn't care to put the effort in to read mine - no, you didn't address all of it, because you ignored the premise and context of having something *native*, and you ignored doveq's question in attempting to respond to it because your solution bypasses what ActiveModel provides, which while out of *explicit* scope of what I said, would still result in support within it at least to some degree.
LLMs can't read subtext. Stop letting LLMs read and write for you, they're bullshitting you and you're generating toxic waste as a result.

I wish you'd realise that using an AI to respond to people strips all trust that you actually care about the problem. Be better. It's slop whether or not you want to acknowledge it as such, it sucks to read.

I'll judge it on what's wrong when you spend more than 5 seconds caring about if something is wrong in it. I'm not going to put in half an hour of work for your 30 seconds of non-effort.
